# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/compiler/translator/AtomicCounterFunctionHLSL.cpp
# Commit: 3a91885d47e0
# Full Hash: 3a91885d47e0ef7fba9749562d52300eea67c651
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-11-11 21:52:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 77cee77c0307 -r 3a91885d47e0 gfx/angle/checkout/src/compiler/translator/AtomicCounterFunctionHLSL.cpp
--- a/gfx/angle/checkout/src/compiler/translator/AtomicCounterFunctionHLSL.cpp	Fri Nov 08 16:41:42 2019 +0000
+++ b/gfx/angle/checkout/src/compiler/translator/AtomicCounterFunctionHLSL.cpp	Fri Nov 08 17:06:00 2019 +0000
@@ -9,6 +9,7 @@
 
 #include "compiler/translator/AtomicCounterFunctionHLSL.h"
 
+#include "compiler/translator/Common.h"
 #include "compiler/translator/ImmutableStringBuilder.h"
 #include "compiler/translator/InfoSink.h"
 #include "compiler/translator/IntermNode.h"
@@ -24,6 +25,10 @@
 constexpr ImmutableString kAtomicCounterBaseName("_acbase_");
 }  // namespace
 
+AtomicCounterFunctionHLSL::AtomicCounterFunctionHLSL(bool forceResolution)
+    : mForceResolution(forceResolution)
+{}
+
 ImmutableString AtomicCounterFunctionHLSL::useAtomicCounterFunction(const ImmutableString &name)
 {
     // The largest string that will be create created is "_acbase_increment" or "_acbase_decrement"
@@ -65,34 +70,41 @@
     {
         out << "uint " << atomicFunction.first
             << "(in RWByteAddressBuffer counter, int address)\n"
-               "{\n";
+               "{\n"
+               "    uint ret;\n";
+
         switch (atomicFunction.second)
         {
             case AtomicCounterFunction::INCREMENT:
+                out << "    counter.InterlockedAdd(address, 1u, ret);\n";
+                break;
             case AtomicCounterFunction::DECREMENT:
-                out << "    uint ret;\n"
-                       "    counter.InterlockedAdd(address, ";
-                if (atomicFunction.second == AtomicCounterFunction::DECREMENT)
-                {
-                    out << "0u - ";
-                }
-                out << "1u, ret);\n"
-                    << "    return ret;\n";
+                out << "    counter.InterlockedAdd(address, 0u - 1u, ret);\n"
+                       "    ret -= 1u;\n";  // atomicCounterDecrement is a post-decrement op
                 break;
             case AtomicCounterFunction::LOAD:
-                out << "    return counter.Load(address);\n";
+                out << "    ret = counter.Load(address);\n";
                 break;
             default:
                 UNREACHABLE();
                 break;
         }
-        out << "}\n\n";
+
+        if (mForceResolution && atomicFunction.second != AtomicCounterFunction::LOAD)
+        {
+            out << "    if (ret == 0) {\n"
+                   "        ret = 0 - ret;\n"
+                   "    }\n";
+        }
+
+        out << "    return ret;\n"
+               "}\n\n";
     }
 }
 
 ImmutableString getAtomicCounterNameForBinding(int binding)
 {
-    std::stringstream counterName;
+    std::stringstream counterName = sh::InitializeStream<std::stringstream>();
     counterName << kAtomicCounterBaseName << binding;
     return ImmutableString(counterName.str());
 }